Sources/SwiftSyntax/WeakLookupTable.swift

```diff
@@ -0,0 +1,213 @@
+  var id: Identifier { get }
+}
+
+private struct WeakReference<T: AnyObject>: ExpressibleByNilLiteral {
+  weak var value: T?
+  init(_ value: T?) {
+    self.value = value
+  }
+  init(nilLiteral: ()) {
+    self.value = nil
+  }
+}
+
+/// 'Set' like container providing lookup table for objects with identifier.
+/// This doesn't take ownership of the objects. Instead, the objects are
+/// held by weak references.
+/// 
+/// References are stored in a hash table with simple open adressing. Because
+/// of weak reference, unlike normal open addressing, erased bucket are simply
+/// turned into 'nil'.
+class WeakLookupTable<Element: Identifiable & AnyObject> {
+
+  /// Storage for the hash table.
+  private var buckets: UnsafeMutablePointer<WeakReference<Element>>
+  private var bucketCount: Int
+
+  /// Estimated count of inserted values. This is greater than or equal to
+  /// the number of actually occupied buckets.
+  /// i.e. estimatedCount >= _countOccupiedBuckets()
+  private var estimatedCount: Int
+
+  init(capacity: Int = 0) {
+    bucketCount = WeakLookupTable<Element>._bucketCount(for: capacity)
+    buckets = .allocate(capacity: bucketCount)
+    buckets.initialize(repeating: nil, count: bucketCount)
+    estimatedCount = 0
+  }
+
+  deinit {
+    buckets.deinitialize(count: bucketCount)
+    buckets.deallocate()
+  }
+
+  /// Constant max load factor for hash table.
+  private static var maxLoadFactor: Double {
+    @inline(__always) get {
+      return 0.75
+    }
+  }
+
+  /// Minimal number of bucket count enough to hold specified capacity of
+  /// objects with taking max load factor into account.
+  private static func _minimalBucketCount(for capacity: Int) -> Int {
+    return Int((Double(capacity) / maxLoadFactor).rounded(.up))
+  }
+
+  /// Number of bucket count to allocate to hold specified number of objects.
+  /// This is the next power of 2 greater than or equal to
+  /// '_minimalBucketCount(for: capacity)'
+  private static func _bucketCount(for capacity: Int,
+                                   from current: Int = 2) -> Int {
+    // Make sure it's representable.
+    precondition(capacity <= (Int.max >> 1) + 1)
+    // Bucket count must always be power of 2.
+    precondition((current & (current - 1)) == 0)
+    // Minimum is 2 to guarantee at least 1 hole.
+    precondition(current >= 2)
+
+    let minimalBucketCount = _minimalBucketCount(for: capacity)
+    var bucketCount = current
+    while bucketCount < minimalBucketCount {
+      bucketCount &*= 2
+    }
+    return bucketCount
+  }
+
+  private var _bucketMask: Int {
+    @inline(__always) get {
+      return bucketCount &- 1
+    }
+  }
+
+  @inline(__always)
+  private func _idealBucket(for id: Element.Identifier) -> Int {
+    return id.hashValue & _bucketMask
+  }
+
+  /// Finds the bucket where the object with the specified id should be stored
+  /// to.
+  private func _findHole(_ id: Element.Identifier) -> (pos: Int, found: Bool) {
```

Comment:
I think `alreadyExists` would be clearer in the return type than `found`. `found` implies that we found a hole which we will always do.